core/device.cpp

```diff
@@ -197,3 +199,102 @@ extern "C" void free_device_table(struct device_table *devices)
 {
 	delete devices;
 }
+
+// managing fingerprint data
+bool fingerprint_record::operator<(const fingerprint_record &a) const
+{
+	if (model == a.model)
+		return serial < a.serial;
+	return model < a.model;
+}
+
+// annoyingly, the Cressi Edy doesn't support a serial number (it's always 0), but still uses fingerprints
+// so we can't bail on the serial number being 0
+extern "C" unsigned int get_fingerprint_data(const struct fingerprint_table *table, uint32_t model, uint32_t serial, unsigned char **fp_out)
+{
+	if (model == 0 || fp_out == nullptr)
+		return 0;
+	struct fingerprint_record fpr = { model, serial };
+	auto it = std::lower_bound(table->fingerprints.begin(), table->fingerprints.end(), fpr);
+	if (it != table->fingerprints.end()) {
+		if (has_dive(it->fdeviceid, it->fdiveid)) {
+			*fp_out = it->raw_data;
+			return it->fsize;
+		}
+	}
+	return 0;
+}
+
+extern "C" void create_fingerprint_node(struct fingerprint_table *table, uint32_t model, uint32_t serial,
+				       const unsigned char *raw_data_in, unsigned int fsize, uint32_t fdeviceid, uint32_t fdiveid)
+{
+	// since raw data can contain \0 we copy this manually, not as string
+	unsigned char *raw_data = (unsigned char *)malloc(fsize);
```

Comment:
At a quick glance, I did not see a matching `free()`.

If you really want to bite the bullet and implement C++-style ownership semantics, you could use a `std::vector<unsigned char>` and instead of `malloc` do a `raw_data.resize(fsize)`? Get a pointer to the data with `&raw_data[0]`.

Or perhaps this is all unnecessary since these things are permanent during the application's life-time.

Reply:
the free needs to happen below where we update the value - if we ever update the value. Otherwise it's indeed app-life-time.
And of course the update is completely bogus as once again it misunderstands what `std::lower_bound` does